### DISCUSSION TITLE
Deploy Azure resources via pipelines

### DISCUSSION THREAD

**Discussion by fgheysels** [arcus-azure/arcus#211] [Ideas]
We're currently in a situation where some of our pipelines fail, because tests that have a dependency on certain Azure resources fail as those Azure resources have been removed.

Why don't we extend our pipelines so that all Azure resources that we're dependent on, are created by those pipelines.  We can do that via bicep templates.

**Comment by stijnmoreels** [comment]
There are cases where we already do this fully/partially. Sometimes it's a drawback and a compromise. Like, if the resource takes a long time before it's fully functional. An example is our SQL server that is used in the scripting library. The server itself already exists on the Azure subscription, but the tables/indexes/... are created 'on the fly'. This approach is used in other libraries as well, reducing the build-time and still be as less dependent as possible.

**Reply by fgheysels** [reply]
It is true that it can take a while to get a resource deployed.  However, once the resource is already there (and doesn't need any modifications), it takes way less time for the deployment to complete.

**Reply by stijnmoreels** [reply]
That, I don't understand 😅 . Is there a difference between deployment and 'already there'?

**Reply by fgheysels** [reply]
Sorry, lost track of this one :)
Yes, therer is a difference.   When deploying Azure resources via ARM or bicep, you can specify the 'deployment method'.  In almost all cases, you specify this as 'incremental'.  This means that you'll execute the template which describes how the Azure resource should look like.  When the requested resource is already there in the requested state, nothing will change.   If the resource is there, but some properties are not as described as in the bicep/ARM template, the resource will be modified.  If the resource is not there, it will be created.

True, it takes some time to execute as it is another task that must be performed.  When the resource is not there yet, it will take longer to execute that step as opposed to when the resource is already there.

**Reply by stijnmoreels** [reply]
We could see if we could make bicep/ARM templates for the necessary Azure resources, but we should also be able to deploy those from our test suite so that the result is always the same. Not sure if that's possible, needs to be investigated.
